capstone-rs/src/instruction.rs

```diff
@@ -203,66 +306,45 @@ pub struct Insn<'a> {
 ///
 pub struct InsnDetail<'a>(pub(crate) &'a cs_detail, pub(crate) Arch);
 
-impl<'a> Insn<'a> {
-    /// Create an `Insn` from a raw pointer to a [`capstone_sys::cs_insn`].
-    ///
-    /// This function serves to allow integration with libraries which generate `capstone_sys::cs_insn`'s internally.
-    ///
-    /// # Safety
-    ///
-    /// Note that this function is unsafe, and assumes that you know what you are doing. In
-    /// particular, it generates a lifetime for the `Insn` from nothing, and that lifetime is in
-    /// no-way actually tied to the cs_insn itself. It is the responsibility of the caller to
-    /// ensure that the resulting `Insn` lives only as long as the `cs_insn`. This function
-    /// assumes that the pointer passed is non-null and a valid `cs_insn` pointer.
-    pub unsafe fn from_raw(insn: *const cs_insn) -> Self {
-        Self {
-            insn: *insn,
-            _marker: PhantomData,
-        }
-    }
-
-    /// The mnemonic for the instruction
-    pub fn mnemonic(&self) -> Option<&str> {
-        unsafe { str_from_cstr_ptr(self.insn.mnemonic.as_ptr()) }
-    }
-
-    /// The operand string associated with the instruction
-    pub fn op_str(&self) -> Option<&str> {
-        unsafe { str_from_cstr_ptr(self.insn.op_str.as_ptr()) }
-    }
-
-    /// Access instruction id
-    pub fn id(&self) -> InsnId {
-        InsnId(self.insn.id)
-    }
-
-    /// Size of instruction (in bytes)
-    fn len(&self) -> usize {
-        self.insn.size as usize
+impl<'a> Drop for Insn<'a> {
```

Comment:
This impl should not be needed since `impl<'a> Drop for Instructions<'a>` handles this.

Reply:
Good catch; it could actually be interesting to use this drop impl to write a `cfg(debug_assertions)` impl of `Drop` that calls `unreachable!()`

Reply:
Interesting idea--but I don't think we want to implement `Drop` because it would affect some borrowing ergonomics because of the [drop check](https://doc.rust-lang.org/nomicon/dropck.html).

Reply:
Precisely because it represents unreachable code, drop check should affect no actual code whatsoever. Still, I reckon that a `cfg`-gated `unreachable!` inside a drop impl seems to push the debug assertions thing a bit too far 😅